File: Data Structures/Stacks and Queues/StackThroughLinkedLists.js
// Stack implementation using a linked list
// LIFO

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.top = null;
        this.bottom = null;
        this.length = 0;
    }

    // See the very top node
    peek(){
        return this.top;
    }

    // Add a node to the top of the stck
    push(value){
        const newNode = new Node(value);
        if (this.length === 0) {
            this.top = newNode;
            this.bottom = newNode;
        } else {
            const holdingPointer = this.top;
            this.top = newNode
            this.top['next'] = holdingPointer
        }
        this.length++;
        return this;
    }

    // Remove the top node from the stack
    pop(){
        if (!this.top) {
            return null;
        }
        if (this.top === this.bottom) {
            this.bottom = null;
        }
        this.top = this.top.next;
        this.length--;
        return this;
    }

    // Check if it it empty
    isEmpty() {
    if (!this.bottom) {
        return true; }
    return false;
    }
}


// Print outs
const myStack = new Stack();
myStack.push('google')
myStack.push('udemy')
myStack.push('youtube')
myStack.peek()
myStack.isEmpty()
myStack.pop()
myStack.pop()
myStack.pop()
myStack.isEmpty()

File: Data Structures/Stacks and Queues/StackThroughLinkedLists.py
# Stack implementation using a linked list
# LIFO

class Stack:
    def __init__(self):
        self.top = None
        self.bottom = None
        self.length = 0
    

    # See the very top node
    def peek(self):
        return self.top


    # Add node to the top of the stck
    def push(self, value):
        new_node = {
            'value': value,
            'next': None
        }
        if self.is_empty():
            self.bottom = new_node
            self.top = new_node
        else:
            holding_pointer = self.top
            self.top = new_node
            self.top['next'] = holding_pointer
        self.length += 1
        return self


    # Remove the top node from the stack
    def pop(self):
        if not self.top:
            return None
        elif self.top == self.bottom:
            self.bottom = None

        else:
            self.top = self.top['next']
        self.length-=1
        return self

    # Check if it it empty
    def is_empty(self):
        if not self.bottom:
            return True
        return False

    

# Print outs
my_stack = Stack()
my_stack.push('google')
my_stack.push('udemy')
my_stack.push('youtube')
print(my_stack.is_empty())
print(my_stack.peek())
my_stack.pop()
my_stack.pop()
my_stack.pop()
print(my_stack.peek())
print(my_stack.is_empty())

